### PR TITLE
Lru cache implementation, needed for things like DNS cache.

### PR DESCRIPTION
**Pull request by JonathanHenson**
Question, should I add a TTL to the cache here? It could be useful, but maybe simplicity dictates that we just do that where we need it.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by RandyGaul** [comment]
Gut says add it when/where needed, and not now.

File: include/aws/common/lru_cache.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/hash_table.h>
+#include <aws/common/linked_list.h>
+
+/**
+ * Simple Least-recently-used cache using the standard lazy linked hash table implementation. (Yes the one that was the
+ * answer to that stupid interview question that one time).
```
[the remark this answers is not in the document]
Comment (JonathanHenson) [reply include/aws/common/lru_cache.h]: done

File: include/aws/common/lru_cache.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ */
+struct aws_lru_cache {
+    struct aws_allocator *allocator;
+    struct aws_linked_list list;
+    struct aws_hash_table table;
+    aws_hash_element_destroy_t user_on_destroy;
```
[the remark this answers is not in the document]
Comment (JonathanHenson) [reply include/aws/common/lru_cache.h]: changed to user_on_value_destroy

File: source/lru_cache.c [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+void aws_lru_cache_clean_up(struct aws_lru_cache *cache) {
+    /* clearing the table will remove all elements. That will also deallocate any cache entries we currently have. */
+    aws_hash_table_clean_up(&cache->table);
+    AWS_ZERO_STRUCT(*cache);
+}
+
+int aws_lru_cache_find(struct aws_lru_cache *cache, const void *key, void **p_value) {
+
+    *p_value = NULL;
```
[the remark this answers is not in the document]
Comment (JonathanHenson) [reply source/lru_cache.c]: good catch

File: source/lru_cache.c [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    void *value;
+};
+
+static void element_destroy(void *value) {
+    struct cache_node *cache_node = value;
+
+    aws_linked_list_remove(&cache_node->node);
+
+    if (cache_node->cache->user_on_destroy) {
+        cache_node->cache->user_on_destroy(cache_node->value);
```
[the remark this answers is not in the document]
Comment (JonathanHenson) [reply source/lru_cache.c]: done

File: include/aws/common/lru_cache.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                                      aws_hash_element_destroy_t destroy_value_fn,
+                                      size_t max_items);
+
+/**
+ * Cleans up the cache. Elements in the cache will be evicted and cleanup callbacks will be invoked.
+ */
+AWS_COMMON_API void aws_lru_cache_clean_up(struct aws_lru_cache *cache);
+
+/**
+ * Finds element in the cache be key. If found, it will become most-recently used, *p_value will hold the stored value,
```
Comment (singku) [inline include/aws/common/lru_cache.h]: by key.

**Pull request merged**